File: day15/main.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	buf, err := io.ReadAll(file)
	if err != nil {
		log.Fatalln(err)
	}
	buf = bytes.TrimSpace(buf)
	words := bytes.Split(buf, []byte{','})

	var boxes [256]Box
	sum := 0
	for _, i := range words {
		sum += hashWord(i)
		processInstr(i, boxes[:])
	}
	fmt.Println("Part 1:", sum)

	sum = 0
	for n, i := range boxes {
		for k, j := range i.values {
			sum += (n + 1) * (k + 1) * j.v
		}
	}
	fmt.Println("Part 2:", sum)
}

type (
	Box struct {
		values []BoxValue
	}

	BoxValue struct {
		id string
		v  int
	}
)

func processInstr(v []byte, m []Box) {
	if r, ok := bytes.CutSuffix(v, []byte{'-'}); ok {
		h := hashWord(r)
		label := string(r)
		for n, i := range m[h].values {
			if i.id == label {
				copy(m[h].values[n:], m[h].values[n+1:])
				m[h].values = m[h].values[:len(m[h].values)-1]
				return
			}
		}
		return
	}
	if r, b, ok := bytes.Cut(v, []byte{'='}); ok {
		num, err := strconv.Atoi(string(b))
		if err != nil {
			log.Fatalln(err)
		}
		h := hashWord(r)
		label := string(r)
		for n, i := range m[h].values {
			if i.id == label {
				m[h].values[n] = BoxValue{
					id: label,
					v:  num,
				}
				return
			}
		}
		m[h].values = append(m[h].values, BoxValue{
			id: label,
			v:  num,
		})
		return
	}
	log.Fatalln("Malformed op")
}

func hashWord(v []byte) int {
	currentValue := 0
	for _, i := range v {
		currentValue = hashStep(currentValue, int(i))
	}
	return currentValue
}

func hashStep(v, b int) int {
	return ((v + b) * 17) % 256
}

File: day15/src/main.rs
use std::fs::File;
use std::io::prelude::*;

const PUZZLEINPUT: &str = "input.txt";

struct BoxValue {
    id: Vec<u8>,
    v: u8,
}

struct LightBox {
    values: Vec<BoxValue>,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut file = File::open(PUZZLEINPUT)?;

    let mut buf = Vec::new();
    file.read_to_end(&mut buf)?;
    let left = buf
        .iter()
        .position(|v| !v.is_ascii_whitespace())
        .unwrap_or(buf.len());
    let right = buf
        .iter()
        .rposition(|v| !v.is_ascii_whitespace())
        .unwrap_or(buf.len());
    let buf = &buf[left..=right];

    let mut boxes: [LightBox; 256] = std::array::from_fn(|_| LightBox { values: Vec::new() });
    let mut sum = 0;
    for i in buf.split(|&v| v == b',') {
        sum += hash_word(i) as usize;
        process_instr(&i, &mut boxes)?;
    }

    println!(
        "Part 1: {}\nPart 2: {}",
        sum,
        boxes.iter().enumerate().fold(0, |acc, (n, i)| acc
            + (n + 1)
                * i.values
                    .iter()
                    .enumerate()
                    .fold(0, |acc, (k, j)| acc + (k + 1) * (j.v as usize)))
    );

    Ok(())
}

fn process_instr(
    instr: &[u8],
    boxes: &mut [LightBox; 256],
) -> Result<(), Box<dyn std::error::Error>> {
    if let Some(label) = instr.strip_suffix(&[b'-']) {
        let h = hash_word(label);
        let values = &mut boxes[h as usize].values;
        if let Some(n) = values.iter().position(|v| v.id == label) {
            values.remove(n);
        }
        return Ok(());
    }
    if let Some(n) = instr.iter().position(|&v| v == b'=') {
        let (label, b) = instr.split_at(n);
        let num = std::str::from_utf8(&b[1..])?.parse::<u8>()?;
        let h = hash_word(label);
        let values = &mut boxes[h as usize].values;
        if let Some(n) = values.iter().position(|v| v.id == label) {
            values[n].v = num;
        } else {
            values.push(BoxValue {
                id: label.to_vec(),
                v: num,
            });
        }
        return Ok(());
    }
    Err("Invalid instr".into())
}

fn hash_word(v: &[u8]) -> u8 {
    v.iter().fold(0, |acc, &v| hash_step(acc, v))
}

fn hash_step(v: u8, b: u8) -> u8 {
    (((v as u32 + b as u32) * 17) % 256) as u8
}
